### PR TITLE
Revised support for Solaris/OmniOS

### PR DESCRIPTION
**Pull request by westbywest**
This PR includes ordering for enabling the following services (in this order):
- nfs/status
- nfs/nlockmgr
- nfs/client

Starting these services in any other order causes nfs/nlockmgr or nfs/client to fail startup on missing dependencies.  Tested on OmniOS r151014 under vagrant.

### CODE REVIEW

**Comment by westbywest** [comment]
Also, inability to enable multiple services simultaneously via svcadm tool appears to be a known bug in Illumos/OmniOS, perhaps not for Solaris itself:
https://www.illumos.org/issues/5593

File: manifests/init.pp [hunk 1]
```diff
@@ -34,6 +34,7 @@
           fail("nfs module only supports lsbdistid Debian and Ubuntu of osfamily Debian. Detected lsbdistid is <${::lsbdistid}>.")
         }
       }
+      $default_nfs_service_required_svcs = undef
```
Comment (ghoneycutt) [inline manifests/init.pp]: why is this here?

File: manifests/init.pp [hunk 2]
```diff
@@ -58,28 +59,33 @@
           fail("nfs module only supports EL 5, 6 and 7 and operatingsystemmajrelease was detected as <${::operatingsystemmajrelease}>.")
         }
       }
+      $default_nfs_service_required_svcs = undef
```
Comment (ghoneycutt) [inline manifests/init.pp]: ?

File: manifests/init.pp [hunk 3]
```diff
@@ -58,28 +59,33 @@
...
     'Solaris': {
       case $::kernelrelease {
         '5.10': {
           $default_nfs_package = [ 'SUNWnfsckr',
-                                    'SUNWnfscr',
-                                    'SUNWnfscu',
-                                    'SUNWnfsskr',
-                                    'SUNWnfssr',
-                                    'SUNWnfssu',
+                                   'SUNWnfscr',
```
Comment (ghoneycutt) [inline manifests/init.pp]: This breaks puppet-lint

ideal style for this is like

https://github.com/ghoneycutt/puppet-module-nfs/blob/omnios_511/manifests/init.pp#L65

File: manifests/init.pp [hunk 4]
```diff
@@ -58,28 +59,33 @@
...
+                                   'SUNWnfscu',
+                                   'SUNWnfsskr',
+                                   'SUNWnfssr',
+                                   'SUNWnfssu',
           ]
+          $default_nfs_service_required_svcs = undef
```
Comment (ghoneycutt) [inline manifests/init.pp]: It seems we have removed the nfs/client service from solaris 10 and 11

File: manifests/init.pp [hunk 5]
```diff
@@ -115,14 +122,24 @@
 
   if $nfs_service == 'USE_DEFAULTS' {
     $nfs_service_real = $default_nfs_service
+    $nfs_service_required_svcs_real = $default_nfs_service_required_svcs
   } else {
     $nfs_service_real = $nfs_service
+    $nfs_service_required_svcs_real = undef
```
Comment (ghoneycutt) [inline manifests/init.pp]: instead of `undef` we should add a parameter for people to specify the services

File: manifests/init.pp [hunk 6]
```diff
@@ -115,14 +122,24 @@
...
   }
 
   package { $nfs_package_real:
     ensure => present,
   }
 
+  if $nfs_service_required_svcs_real and $nfs_service_real {
+   service { $nfs_service_required_svcs_real:
```
Comment (ghoneycutt) [inline manifests/init.pp]: I thought you cant specify an array to svcadm, so I'm confused to see it here.

File: manifests/init.pp [hunk 7]
```diff
@@ -115,14 +122,24 @@
...
   } else {
     $nfs_service_real = $nfs_service
+    $nfs_service_required_svcs_real = undef
   }
 
   package { $nfs_package_real:
     ensure => present,
   }
 
+  if $nfs_service_required_svcs_real and $nfs_service_real {
```
Comment (ghoneycutt) [inline manifests/init.pp]: Better to be explicit

`if $nfs_service_required_svcs_real != undef and $nfs_service_real != undef`

**Comment by ghoneycutt** [comment]
Awesome work. I've got a few questions and this is well on the way to being merged

File: manifests/init.pp (hunk 6 above)
Comment (westbywest) [inline manifests/init.pp]: In Puppet resource declarations, supplying an array in the title attribute has same effect as declaring a separate resource for each item in the array.  I.e. it would invoke svcadm multiple times.  Supplying an array to the "name" attribute apparently only casts the array into a concatenated string that is then supplied to the underlying package provider tool (e.g. svcadm).

File: manifests/init.pp (hunk 1 above)
Comment (westbywest) [inline manifests/init.pp]: Added for completeness, i.e. so this new variable would have a value assigned (even if undef) for each clause of case statement.

File: manifests/init.pp (hunk 2 above)
Comment (westbywest) [inline manifests/init.pp]: See previous comment.

File: manifests/init.pp (hunk 5 above)
Comment (westbywest) [inline manifests/init.pp]: I can expose as parameter for completeness.

File: manifests/init.pp (hunk 2 above)
Comment (ghoneycutt) [inline manifests/init.pp]: aha, gotcha!

**Comment by ghoneycutt** [comment]
Since OmniOS has the bug with multiple services being passed, should we have a separate service resource block for each service?

If that's the case, it might be easier to do a simple conditional like this pseudo-code

``` puppet
if $platform is Solaris or OmniOS {

  service { 'nfs/status':
    ...
  }
  service { 'nfs/nlockmgr':
    ..
  }
}
```

**Comment by westbywest** [comment]
(Repeating a comment on outdated diff above.) In Puppet resource declarations supplying an array in the title attribute has same effect as declaring a separate resource for each item in the array. I.e. it would invoke svcadm multiple times.

I'm assuming that doing multiple resource declarations, where implicitly via array as title attribute or explicitly in your pseudo-code above, would have the same effect on both OmniOS and Solaris.

**Comment by ghoneycutt** [comment]
@westbywest Thanks!

I've added spec tests and tweaked a few things in PR #68
